### PR TITLE
fix #44705, fix `tuple_tfunc` on `Union` containing `Type{...}`

### LINKED ISSUE #44705: tuple_tfunc bug handling of `Union` containing `Type{...}`
I guess this issue should be raised before, but I couldn't find an issue that discusses this.
When working on #44512 (in particularity [this](https://github.com/JuliaLang/julia/pull/44512#issuecomment-1075345754) issue), I found the following `isa` check doesn't work as I expected:
```julia
julia> Type{Int} <: DataType
true

julia> Int isa Type{Int}
true

julia> (Int,) isa Tuple{DataType} # expected
true

julia> (Int,) isa Tuple{Type{Int}} # I expected this to return `true`
false
```

So in essence `Tuple`-type whose parameter contains `Type`-type doesn't seem to follow the covariant rule.

This issue leads to the `ERROR: fatal error in type inference (type bound)` error in #44512 when we try to union-split `Base.indexed_iterate(::Union{Tuple{Nothing, Nothing, Nothing}, Tuple{Any, Union{Type{AbstractContinuousSystem}, Type{AbstractDiscreteSystem}}, Any}}, ...)` callsite.

### PR DESCRIPTION
**Pull request by aviatesk**

### CODE REVIEW

File: test/compiler/inference.jl [hunk 1]
```diff
@@ -1562,6 +1562,12 @@ end
 @test arraysize_tfunc(Vector, Float64) === Union{}
 @test arraysize_tfunc(String, Int) === Union{}
 
+# tuple_tfunc
+# https://github.com/JuliaLang/julia/issues/44705
+import Core.Compiler: tuple_tfunc, widenconst
+@test widenconst(tuple_tfunc(Any[Type{Int}])) === Tuple{DataType}
+@test tuple_tfunc(Any[Union{Type{Int32},Type{Int64}}]) === Tuple{Type}
```
Comment (Keno) [inline test/compiler/inference.jl:1569]: Why not `Tuple{DataType}`?

File: base/compiler/tfuncs.jl [hunk 2]
```diff
@@ -1554,6 +1558,8 @@ function tuple_tfunc(argtypes::Vector{Any})
                 else
                     params[i] = Type
                 end
+            elseif !isvarargtype(x) && x <: Type
```
Comment (vtjnash) [inline base/compiler/tfuncs.jl:1561]: You need to check hasintersect not subtyping here, then can use Union with Type to widen it successfully

File: test/compiler/inference.jl (hunk 1 above)
Comment (vtjnash) [reply test/compiler/inference.jl:1569]: We would need a more precise idea of widenconst that does widenconcrete, and we don't have that right now

File: base/compiler/tfuncs.jl (hunk 2 above)
Comment (aviatesk) [reply base/compiler/tfuncs.jl:1561]: I found `hasintersect(x, Type)` here is too permissive as it holds for e.g. `x = Any`.
But I guess it's okay if we do `params[i] = Union{x, Type}`?

**Comment by aviatesk** [comment]
`buildbot/package_xxx` builds started to fail somehow, although I can't really reproduce it locally..

File: base/compiler/tfuncs.jl [hunk 3]
```diff
@@ -1554,6 +1558,8 @@ function tuple_tfunc(argtypes::Vector{Any})
                 else
                     params[i] = Type
                 end
+            elseif !isvarargtype(x) && hasintersect(x, Type)
```
Comment (aviatesk) [inline base/compiler/tfuncs.jl:1561]: I encountered a weird behavior here -- without this `!isvarargtype(x)` check `buildbot/xxx` builds fail with the following error:
```
Assertion failed: (!jl_is_vararg(x) && !jl_is_vararg(y)), function subtype, file /Users/julia/buildbot/worker/package_macos64/build/src/subtype.c, line 1261.
```
(e.g. https://build.julialang.org/#/builders/34/builds/4697)
while the same commit successfully builds w/o any problem on my machine (after `make cleanall` or even a fresh clone).
Maybe buildbot uses a different method to build something, and in that case the bootstrap makes more use of interpreter or something?

Comment (aviatesk) [reply base/compiler/tfuncs.jl:1561]: Turned out that I needed to enable ["assert build"](https://github.com/JuliaLang/julia/blob/master/doc/src/devdocs/build/build.md#building-an-assert-build-of-julia).

**Comment by aviatesk** [comment]
This causes `SquadGames.jl` test suite to fail.
Here is the reproducer:
```julia
function assertx(xs::Vector{T}) where T
    t = (xs[1], xs[2])
    t::Tuple{T,T}
end

@code_typed optimize=false assertx(Type{<:Number}[Int,Int])
@code_typed optimize=true assertx(Type{<:Number}[Int,Int])
```

```diff
diff --git a/_master.jl b/_pr.jl
index 03768efc773..54ccceaa2cc 100644
--- a/_master.jl
+++ b/_pr.jl
@@ -2,8 +2,8 @@ julia> @code_typed optimize=false assertx(Type{<:Number}[Int,Int])
 CodeInfo(
 1 ─ %1 = Base.getindex(xs, 1)::Type{<:Number}
 │   %2 = Base.getindex(xs, 2)::Type{<:Number}
-│        (t = Core.tuple(%1, %2))::Tuple{Type{<:Number}, Type{<:Number}}
-│   %4 = t::Tuple{Type{<:Number}, Type{<:Number}}
+│        (t = Core.tuple(%1, %2))::Tuple{Type, Type}
+│   %4 = t::Tuple{Type, Type}
 │   %5 = Core.apply_type(Main.Tuple, $(Expr(:static_parameter, 1)), $(Expr(:static_parameter, 1)))::Core.Const(Tuple{Type{<:Number}, Type{<:Number}})
 │   %6 = Core.typeassert(%4, %5)::Tuple{Type{<:Number}, Type{<:Number}}
 └──      return %6
@@ -13,6 +13,8 @@ julia> @code_typed optimize=true assertx(Type{<:Number}[Int,Int])
 CodeInfo(
 1 ─ %1 = Base.arrayref(true, xs, 1)::Type{<:Number}
 │   %2 = Base.arrayref(true, xs, 2)::Type{<:Number}
-│   %3 = Core.tuple(%1, %2)::Tuple{Type{<:Number}, Type{<:Number}}
-└──      return %3
+│   %3 = Core.tuple(%1, %2)::Tuple{Type, Type}
+│        Core.typeassert(%3, Tuple{Type{<:Number}, Type{<:Number}})::Tuple{Type{<:Number}, Type{<:Number}}
+│   %5 = π (%3, Tuple{Type{<:Number}, Type{<:Number}})
+└──      return %5
 ) => Tuple{Type{<:Number}, Type{<:Number}}
```

So it seems like this assertion error was wrongly erased before? I wonder if we want to disable constructing a container whose element type is `Tuple{Type{...}}`.

**Pull request merged**